pytensor/tensor/rewriting/elemwise.py

```diff
@@ -529,333 +539,486 @@ def local_upcast_elemwise_constant_inputs(fgraph, node):
-                assert len(ret) == len(new_node.outputs)
-                assert len(ret) == 1
-                new_node = ret[0].owner
-            else:
-                break
+    def apply(self, fgraph):
+        nb_replacement = 0
 
-        return new_node.outputs
+        if fgraph.profile:
+            validate_before = fgraph.profile.validate_time
+            callbacks_before = fgraph.execute_callbacks_times.copy()
+            callback_before = fgraph.execute_callbacks_time
 
-    return local_fuse
+        max_operands = elemwise_max_operands_fct(None)
+
+        def find_next_fuseable_subgraph(
+            fg: FunctionGraph,
+        ) -> Generator[Tuple[List[Variable], List[Variable]], None, None]:
+            """Find all subgraphs in a FunctionGraph that can be fused together
+
+            Yields
+            -------
+            List of inputs and outputs that determine subgraphs which can be fused. This
+            method assumes that such replacement is done across iterations of the
+            generator.
+            """
+
+            @lru_cache(maxsize=None)
+            def elemwise_scalar_op_has_c_code(node: Apply) -> bool:
+                # TODO: This should not play a role in non-c backends!
+                if node.op.scalar_op.supports_c_code(node.inputs, node.outputs):
+                    return True
+                else:
+                    warn(
+                        (
+                            "Optimization Warning: "
+                            f"The Op {node.op.scalar_op} does not provide a C implementation."
+                            " As well as being potentially slow, this also disables "
+                            "loop fusion."
+                        )
+                    )
+                    return False
+
+            # We start by creating two maps, 1) from each node to each potentially
+            # fuseable client (both nodes must be single output Elemwise with same
+            # broadcast type) and 2) from each node to each certainly unfuseable
+            # client (those that don't fit into 1))
+            fuseable_clients: Dict[Any, List[Any]] = defaultdict(list)
+            unfuseable_clients: Dict[Any, List[Any]] = defaultdict(list)
+            for out, clients in fg.clients.items():
+                out_maybe_fuseable = (
+                    out.owner
+                    and isinstance(out.owner.op, Elemwise)
+                    # and not isinstance(out.owner.op.scalar_op, aes.Composite)
+                    and len(out.owner.outputs) == 1
+                    and elemwise_scalar_op_has_c_code(out.owner)
+                )
+                for client, _ in clients:
+                    if (
+                        out_maybe_fuseable
+                        and not isinstance(client, str)  # "output"
+                        and isinstance(client.op, Elemwise)
+                        # and not isinstance(client.op.scalar_op, aes.Composite)
```

Comment:
todo

Reply:
This is not a TODO, it's a question of whether we want to combine them or not. For now I decided not to